crates/services/gas_price_service/src/v1/service.rs

```diff
@@ -136,63 +139,79 @@ where
     }
 
     fn validate_block_gas_capacity(
-        &self,
         block_gas_capacity: u64,
     ) -> anyhow::Result<NonZeroU64> {
         NonZeroU64::new(block_gas_capacity)
             .ok_or_else(|| anyhow!("Block gas capacity must be non-zero"))
     }
 
-    async fn set_metadata(&mut self) -> anyhow::Result<()> {
-        let metadata: UpdaterMetadata = self.algorithm_updater.clone().into();
-        self.metadata_storage
-            .set_metadata(&metadata)
-            .map_err(|err| anyhow!(err))
-    }
-
-    async fn handle_normal_block(
-        &mut self,
+    async fn handle_normal_block<'a>(
+        &'a mut self,
         height: u32,
         gas_used: u64,
         block_gas_capacity: u64,
         block_bytes: u64,
         block_fees: u64,
-    ) -> anyhow::Result<()> {
-        let capacity = self.validate_block_gas_capacity(block_gas_capacity)?;
+    ) -> anyhow::Result<()>
+    where
+        StorageTxProvider::Transaction<'a>:
+            UnrecordedBlocks + SetMetadataStorage + SetDaSequenceNumber,
+    {
+        let mut storage_tx = self.storage_tx_provider.begin_transaction()?;
```

Comment:
We might want to start the transaction after we validate the block below. The validation itself is super-cheap, but still it may fail and we'll create the transaction unnecessarily.